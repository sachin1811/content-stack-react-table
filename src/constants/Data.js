export const columnData = [
    {
        Header: "Name",
        accessor: "name"
    },
    {
        Header: "Age",
        accessor: "age"
    },
    {
        Header: "Gender",
        accessor: "gender"
    },
    {
        Header: "Email",
        accessor: "email"
    },
    {
        Header: "Action",
        accessor: "edit"
    }
];

export const EmptyDataFeedback = "No Data Present in Table";

export const mockedData = {
    users: [{ id: 1, firstName: "Sachin", age: "29", email: "email@email.com" }, { id: 2, firstName: "Nitin", age: "33", email: "email@email.com" }]
};
