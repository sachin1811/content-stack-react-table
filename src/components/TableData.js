import { useState,useEffect } from "react";
import ColumnResizer from "react-table-column-resizer";
import { FaEllipsisV } from "react-icons/fa";
import { columnData, EmptyDataFeedback } from "../constants/Data";
import '../styles/style.css';
 

export function TableData({dataPresent=true}) {
    const [data, setData] = useState([]);
    const [offset, setOffset] = useState(0);
    const [isLoading,setIsLoading] = useState(true);



    useEffect(() => {
      if(isLoading)
      {
       fetch(`https://dummyjson.com/users?limit=10&skip=${offset}&select=id,firstName,lastName,age,gender,email`)
       .then(data => data.json())
       .then(resData => resData.users)
       .then(resData => {
        if(!resData.length)
        {
        setIsLoading(false);
        }
        else
        {
        setData([...data,...resData])
       }
       });
      }
    },[offset,dataPresent]);



    const handleScroll = () => {
        const element = document.getElementById('dynamicTable');
        if((element.scrollHeight - Math.round(element.scrollTop) === (element.clientHeight+1)) && isLoading)
        setOffset(x => x + 10);
    }


    

  return (
    <div id="dynamicTable" onScroll={dataPresent?handleScroll:null} className={dataPresent?"border border-solid border-[#DDE3EE] h-[52vh] w-[90vw] overflow-y-scroll m-[5vw] text-xs":"border border-solid border-[#DDE3EE] w-[90vw] m-[5vw] text-xs"}>
      <table className="w-full min-w-max table-auto text-left">
        <thead className="border-b border-[#DDE3EE] sticky top-[-1vh] h-8">
          <tr>
          {columnData.map((column,index) => (
           <><th key={index}
             className="bg-white p-4">
                  {column.Header}
            </th>
            {index!=columnData.length - 1?<ColumnResizer className="max-w-full min-w-0 w-0 bg-[#DDE3EE]" />:null}</>
          ))}
          </tr>
        </thead>
        <tbody>
          {dataPresent?data.map(({ id,firstName,age, gender, email }) => {
            const classes = "p-4";
            const resizerClass = "min-w-0 w-0 bg-[#DDE3EE]";
            return (
              <tr key={id} className="h-12 md:h-15">
                <td className={classes}>
                    {firstName}
                </td>
                <td className={resizerClass}></td>
                <td className={classes}>
                    {age}
                </td>
                <td className={resizerClass}></td>
                <td className={classes}>
                    {gender}
                </td>
                <td className={resizerClass}></td>
                <td className={classes}>
                    {email}
                </td>
                <td className={resizerClass}></td>
                <td className="p-4 w-2 content-center">
                    <FaEllipsisV />
                </td>
              </tr>
            );
          }):( <tr className="h-12 md:h-15">
          <td className="p-4 text-center" colSpan={9}>{EmptyDataFeedback}
          </td></tr>)}
        </tbody>
      </table>
    </div>
  );
}