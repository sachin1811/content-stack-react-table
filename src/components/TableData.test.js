import TestRenderer from 'react-test-renderer';
import { TableData } from './TableData';
import { mockFetch } from '../../_mocks_/mock-fetch';
import { mockedData } from '../constants/Data';

describe("TableData Component",()=>{

    describe("TableData Component without data",()=>{
        let renderer;
        let instance;

        renderer = TestRenderer.create(
            <TableData dataPresent={false} />,
          );

          it("should make snapshot", () => {
            expect(renderer).toMatchSnapshot();
            });
    });

    describe("TableData Component with data",()=>{
        window.fetch = mockFetch(mockedData);
        const mockedFetch = window.fetch;
        let renderer;
        let instance;

        const { act } = TestRenderer;
        renderer = TestRenderer.create(
            <TableData />,
          );
          instance = renderer.root;

          it("should make snapshot",  () => {
          expect(renderer).toMatchSnapshot();
          });

          it("should make snapshot", async () => {
            let treeNode = instance.findByProps({id:"dynamicTable"});
            act(() => {
                treeNode.props.onScroll();
              });
              expect(mockedFetch).toHaveBeenLastCalledWith(true);
            });
    });
});