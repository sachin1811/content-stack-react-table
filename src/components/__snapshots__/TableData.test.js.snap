// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div
  className="border border-solid border-[#DDE3EE] w-[90vw] m-[5vw] text-xs"
  id="dynamicTable"
  onScroll={null}
>
  <table
    className="w-full min-w-max table-auto text-left"
  >
    <thead
      className="border-b border-[#DDE3EE] sticky top-[-1vh] h-8"
    >
      <tr>
        <th
          className="bg-white p-4"
        >
          Name
        </th>
        <th
          className="column_resizer_own_class  max-w-full min-w-0 w-0 bg-blue-gray-100"
          disabled={false}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "cursor": "ew-resize",
              "userSelect": "none",
            }
          }
        />
        <th
          className="bg-white p-4"
        >
          Age
        </th>
        <th
          className="column_resizer_own_class  max-w-full min-w-0 w-0 bg-blue-gray-100"
          disabled={false}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "cursor": "ew-resize",
              "userSelect": "none",
            }
          }
        />
        <th
          className="bg-white p-4"
        >
          Gender
        </th>
        <th
          className="column_resizer_own_class  max-w-full min-w-0 w-0 bg-blue-gray-100"
          disabled={false}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "cursor": "ew-resize",
              "userSelect": "none",
            }
          }
        />
        <th
          className="bg-white p-4"
        >
          Email
        </th>
        <th
          className="column_resizer_own_class  max-w-full min-w-0 w-0 bg-blue-gray-100"
          disabled={false}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "cursor": "ew-resize",
              "userSelect": "none",
            }
          }
        />
        <th
          className="bg-white p-4"
        >
          Action
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="h-12 md:h-15"
      >
        <td
          className="p-4 text-center"
          colSpan={9}
        >
          No Data Present in Table
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[` 2`] = `
<div
  className="border border-solid border-[#DDE3EE] h-[52vh] w-[90vw] overflow-y-scroll m-[5vw] text-xs"
  id="dynamicTable"
  onScroll={[Function]}
>
  <table
    className="w-full min-w-max table-auto text-left"
  >
    <thead
      className="border-b border-[#DDE3EE] sticky top-[-1vh] h-8"
    >
      <tr>
        <th
          className="bg-white p-4"
        >
          Name
        </th>
        <th
          className="column_resizer_own_class  max-w-full min-w-0 w-0 bg-blue-gray-100"
          disabled={false}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "cursor": "ew-resize",
              "userSelect": "none",
            }
          }
        />
        <th
          className="bg-white p-4"
        >
          Age
        </th>
        <th
          className="column_resizer_own_class  max-w-full min-w-0 w-0 bg-blue-gray-100"
          disabled={false}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "cursor": "ew-resize",
              "userSelect": "none",
            }
          }
        />
        <th
          className="bg-white p-4"
        >
          Gender
        </th>
        <th
          className="column_resizer_own_class  max-w-full min-w-0 w-0 bg-blue-gray-100"
          disabled={false}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "cursor": "ew-resize",
              "userSelect": "none",
            }
          }
        />
        <th
          className="bg-white p-4"
        >
          Email
        </th>
        <th
          className="column_resizer_own_class  max-w-full min-w-0 w-0 bg-blue-gray-100"
          disabled={false}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "cursor": "ew-resize",
              "userSelect": "none",
            }
          }
        />
        <th
          className="bg-white p-4"
        >
          Action
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;

exports[`TableData Component TableData Component with data should make snapshot 1`] = `
<div
  className="border border-solid border-[#DDE3EE] h-[52vh] w-[90vw] overflow-y-scroll m-[5vw] text-xs"
  id="dynamicTable"
  onScroll={[Function]}
>
  <table
    className="w-full min-w-max table-auto text-left"
  >
    <thead
      className="border-b border-[#DDE3EE] sticky top-[-1vh] h-8"
    >
      <tr>
        <th
          className="bg-white p-4"
        >
          Name
        </th>
        <th
          className="column_resizer_own_class  max-w-full min-w-0 w-0 bg-[#DDE3EE]"
          disabled={false}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "cursor": "ew-resize",
              "userSelect": "none",
            }
          }
        />
        <th
          className="bg-white p-4"
        >
          Age
        </th>
        <th
          className="column_resizer_own_class  max-w-full min-w-0 w-0 bg-[#DDE3EE]"
          disabled={false}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "cursor": "ew-resize",
              "userSelect": "none",
            }
          }
        />
        <th
          className="bg-white p-4"
        >
          Gender
        </th>
        <th
          className="column_resizer_own_class  max-w-full min-w-0 w-0 bg-[#DDE3EE]"
          disabled={false}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "cursor": "ew-resize",
              "userSelect": "none",
            }
          }
        />
        <th
          className="bg-white p-4"
        >
          Email
        </th>
        <th
          className="column_resizer_own_class  max-w-full min-w-0 w-0 bg-[#DDE3EE]"
          disabled={false}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "cursor": "ew-resize",
              "userSelect": "none",
            }
          }
        />
        <th
          className="bg-white p-4"
        >
          Action
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="h-12 md:h-15"
      >
        <td
          className="p-4"
        >
          Sachin
        </td>
        <td
          className="min-w-0 w-0 bg-[#DDE3EE]"
        />
        <td
          className="p-4"
        >
          29
        </td>
        <td
          className="min-w-0 w-0 bg-[#DDE3EE]"
        />
        <td
          className="p-4"
        />
        <td
          className="min-w-0 w-0 bg-[#DDE3EE]"
        />
        <td
          className="p-4"
        >
          email@email.com
        </td>
        <td
          className="min-w-0 w-0 bg-[#DDE3EE]"
        />
        <td
          className="p-4 w-2 content-center"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 192 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
            />
          </svg>
        </td>
      </tr>
      <tr
        className="h-12 md:h-15"
      >
        <td
          className="p-4"
        >
          Nitin
        </td>
        <td
          className="min-w-0 w-0 bg-[#DDE3EE]"
        />
        <td
          className="p-4"
        >
          33
        </td>
        <td
          className="min-w-0 w-0 bg-[#DDE3EE]"
        />
        <td
          className="p-4"
        />
        <td
          className="min-w-0 w-0 bg-[#DDE3EE]"
        />
        <td
          className="p-4"
        >
          email@email.com
        </td>
        <td
          className="min-w-0 w-0 bg-[#DDE3EE]"
        />
        <td
          className="p-4 w-2 content-center"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 192 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
            />
          </svg>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`TableData Component TableData Component without data should make snapshot 1`] = `
<div
  className="border border-solid border-[#DDE3EE] w-[90vw] m-[5vw] text-xs"
  id="dynamicTable"
  onScroll={null}
>
  <table
    className="w-full min-w-max table-auto text-left"
  >
    <thead
      className="border-b border-[#DDE3EE] sticky top-[-1vh] h-8"
    >
      <tr>
        <th
          className="bg-white p-4"
        >
          Name
        </th>
        <th
          className="column_resizer_own_class  max-w-full min-w-0 w-0 bg-[#DDE3EE]"
          disabled={false}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "cursor": "ew-resize",
              "userSelect": "none",
            }
          }
        />
        <th
          className="bg-white p-4"
        >
          Age
        </th>
        <th
          className="column_resizer_own_class  max-w-full min-w-0 w-0 bg-[#DDE3EE]"
          disabled={false}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "cursor": "ew-resize",
              "userSelect": "none",
            }
          }
        />
        <th
          className="bg-white p-4"
        >
          Gender
        </th>
        <th
          className="column_resizer_own_class  max-w-full min-w-0 w-0 bg-[#DDE3EE]"
          disabled={false}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "cursor": "ew-resize",
              "userSelect": "none",
            }
          }
        />
        <th
          className="bg-white p-4"
        >
          Email
        </th>
        <th
          className="column_resizer_own_class  max-w-full min-w-0 w-0 bg-[#DDE3EE]"
          disabled={false}
          onMouseDown={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "cursor": "ew-resize",
              "userSelect": "none",
            }
          }
        />
        <th
          className="bg-white p-4"
        >
          Action
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className="h-12 md:h-15"
      >
        <td
          className="p-4 text-center"
          colSpan={9}
        >
          No Data Present in Table
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
